cli: sanitize the behavior of `--url` vs other options

knz opened the pull request:
Fixes #23688.
Fixes #22376.
Fixes #23739.

Prior to this patch, if `--url` is specified, that is used to open the
connection regardless of the value of `--database`, `--host`,
`--insecure`, `--port` and `--user`. The value of these flags would
only be used if `--url` is not specified.

This is surprising because if the URL specifies something and the
operator also specifies contradictory information via another
parameter, the other parameter is dropped on the floor without
warning so the operator may get a different connection behavior than
what they expected.

Also if some information is outright missing from the passed URL, it
is not possible to override it using an individual flag (e.g. to set a
current database).

This patch changes this to make `--url` "play nice" with the other
flag: a warning is reported if `--url` is specified and also one of
the other individual flags, to remind the user that `--url`
prevails. Missing items in the URL are "filled in" from other
arguments.

This behavior extends that of `psql`, which always let the URL
prevail, even if it misses information. This patch makes CockroachDB
arguably better, by informing the user about what's going on if
there's a conflict, and enabling some customization of the URL.

Incidentally, this fixes a bug where the `application_name` option was
forcefully introduced in various places where it shouldn't be present.

Release note (cli change): the client commands now reports a
warning if both `--url` is specified and some other command-line
argument that specifies the same information.

Release note (bug fix): the postgres URL reported upon `cockroach
start` again properly does not include the option `application_name`.

knz committed:
cli: sanitize the behavior of `--url` vs other options

Prior to this patch, if `--url` is specified, that is used to open the
connection regardless of the value of `--database`, `--host`,
`--insecure`, `--port` and `--user`. The value of these flags would
only be used if `--url` is not specified.

This is surprising because if the URL specifies something and the
operator also specifies contradictory information via another
parameter, the other parameter is dropped on the floor without
warning so the operator may get a different connection behavior than
what they expected.

Also if some information is outright missing from the passed URL, it
is not possible to override it using an individual flag (e.g. to set a
current database).

This patch changes this to make `--url` "play nice" with the other
flag: a warning is reported if `--url` is specified and also one of
the other individual flags, to remind the user that `--url`
prevails. Missing items in the URL are "filled in" from other
arguments.

This behavior extends that of `psql`, which always let the URL
prevail, even if it misses information. This patch makes CockroachDB
arguably better, by informing the user about what's going on if
there's a conflict, and enabling some customization of the URL.

Incidentally, this fixes a bug where the `application_name` option was
forcefully introduced in various places where it shouldn't be present.

Release note (cli change): the client commands now reports a
warning if both `--url` is specified and some other command-line
argument that specifies the same information.

Release note (bug fix): the postgres URL reported upon `cockroach
start` again properly does not include the option `application_name`.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/23725)
<!-- Reviewable:end -->

bdarnell commented:
Review status: 0 of 4 files reviewed at latest revision, all discussions resolved, some commit checks pending.

---

*[pkg/cli/sql_util.go, line 460 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23725#-L7Q-G6Y9ZFQ5qnM8SkO:-L7Q-G6Y9ZFQ5qnM8SkP:btn01xm) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e7921eb5c7e570e374e167f22a41c875819b21d/pkg/cli/sql_util.go#L460)):*
> ```Go
> 		return nil, err
> 	}
> 	if cliCtx.sqlConnDBName != "" {
> ```

Why only database? I think it would make sense to merge all of the applicable flags into the URL so you can for example copy-paste a URL and then override the `--user`. This would also address #22376. 

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23725)*
<!-- Sent from Reviewable.io -->

knz commented:
I wasn't aware of #22376 and just responded to peter's request that I fix #23688. I'll have a look at it, it seems clear that this can be solved on the same code path. Thanks for the pointer.

knz commented:
Review status: 0 of 9 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[pkg/cli/sql_util.go, line 460 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23725#-L7Q-G6Y9ZFQ5qnM8SkO:-L7R2ncs0Mmgaua_OGAr:b-oq29fq) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e7921eb5c7e570e374e167f22a41c875819b21d/pkg/cli/sql_util.go#L460)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Why only database? I think it would make sense to merge all of the applicable flags into the URL so you can for example copy-paste a URL and then override the `--user`. This would also address #22376. 
</blockquote></details>

Done, RFAL

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23725)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 10 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[pkg/cli/sql_util.go, line 516 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23725#-L7R7JS44_D164WQO6-J:-L7R7JS44_D164WQO6-K:b-9o1dgc) ([raw file](https://github.com/cockroachdb/cockroach/blob/b29c27b66fe2c468aa02db89ee62d715c1d12854/pkg/cli/sql_util.go#L516)):*
> ```Go
> 
> 	// Determine the server to use.
> 	if baseURL.Host != "" && !strings.HasPrefix(baseURL.Host, ":") {
> ```

All this logic seems complex. An alternative way to approach this problem would be to add custom flag values for `--dbname`, `--certs`, etc which update `cliCtx.sqlConnURL`. That is, we'd get rid of `cliCtx.sqlConnDBName`, `cliCtx.sqlConnUser`, etc. An advantage is that the last bit of the command line takes precedence. For example, `--database=foo --url=.../bar`. I'm not sure if this matters in practice and I'm not sure if it would be simpler or a less surprising UX. Consider this a drive-by comment that you can feel free to reject. 

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23725)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Reviewed 12 of 12 files at r2.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[pkg/base/config.go, line 212 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23725#-L7R79TYl6laFkhDiPCR:-L7R79TYl6laFkhDiPCS:b-oz5qhx) ([raw file](https://github.com/cockroachdb/cockroach/blob/b29c27b66fe2c468aa02db89ee62d715c1d12854/pkg/base/config.go#L212)):*
> ```Go
> 
> // LoadSecurityOptions extends a url.Values with SSL settings suitable for
> // the given server config. It returns true if and only of the URL
> ```

s/of/if/

---

*[pkg/cli/sql_util.go, line 471 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23725#-L7R90CWfhKbX6JN_AuJ:-L7R90CWfhKbX6JN_AuK:bxtqoj2) ([raw file](https://github.com/cockroachdb/cockroach/blob/b29c27b66fe2c468aa02db89ee62d715c1d12854/pkg/cli/sql_util.go#L471)):*
> ```Go
> // settings set by the command-line flags. The value of --url, if any,
> // is used as a starting point, then the remaining command-line
> // arguments can override this. There are two exceptions:
> ```

As implemented, the remaining command-line arguments are ignored instead of overriding the URL. I'd prefer that the URL be the lower-priority flag (processing the flags in-order would also be acceptable to me, but I think this would end up being trickier than saying that the other flags always override the URL).

My reasoning is that the URL is likely to be copy/pasted as a single string, so it's nice to give users the option to copy/paste the URL printed at startup (which hard-codes `root@`) and then override it with `--user=me` instead of editing the URL. 

---

*[pkg/cli/sql_util.go, line 472 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23725#-L7R95dvVlgCdddCPbm4:-L7R95dvVlgCdddCPbm5:b-k8cjaz) ([raw file](https://github.com/cockroachdb/cockroach/blob/b29c27b66fe2c468aa02db89ee62d715c1d12854/pkg/cli/sql_util.go#L472)):*
> ```Go
> // is used as a starting point, then the remaining command-line
> // arguments can override this. There are two exceptions:
> // - if the URL already specifies sslmode, then --insecure does not
> ```

Why is this an exception? I think that `--insecure` and `--cert-dir` should be handled in the same way as the other client configuration flags. 

---

*[pkg/cli/user.go, line 41 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23725#-L7R7ZdYaBDh6pzKyOMR:-L7R7ZdYaBDh6pzKyOMS:bxw1iam) ([raw file](https://github.com/cockroachdb/cockroach/blob/b29c27b66fe2c468aa02db89ee62d715c1d12854/pkg/cli/user.go#L41)):*
> ```Go
> 		return usageAndError(cmd)
> 	}
> 	conn, err := getPasswordAndMakeSQLClient("cockroach user get")
> ```

I don't think we want the CLI subcommands to be this fine-grained. We should at least put all the `cockroach user` variants as one "application". Maybe `cockroach sql` gets its own "application" and all the rest are just `cockroach cli`.  

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23725)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[pkg/cli/sql_util.go, line 471 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23725#-L7R90CWfhKbX6JN_AuJ:-L7RbuQ5Hl2sOuYRDhgz:bz4u7by) ([raw file](https://github.com/cockroachdb/cockroach/blob/b29c27b66fe2c468aa02db89ee62d715c1d12854/pkg/cli/sql_util.go#L471)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

As implemented, the remaining command-line arguments are ignored instead of overriding the URL. I'd prefer that the URL be the lower-priority flag (processing the flags in-order would also be acceptable to me, but I think this would end up being trickier than saying that the other flags always override the URL).

My reasoning is that the URL is likely to be copy/pasted as a single string, so it's nice to give users the option to copy/paste the URL printed at startup (which hard-codes `root@`) and then override it with `--user=me` instead of editing the URL. 
</blockquote></details>

Per my other comment, later command line flags overriding earlier ones is a consistent experience. And it has precedent in something like `--database=foo --database=bar`. I think everyone would naturally expect the second flag to override the first. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23725)*
<!-- Sent from Reviewable.io -->

knz commented:
@petermattis after careful consideration, and taking advice from @mberhault, I have decided against doing what you suggested. The main problems are:

- the code to achieve what you suggest is non-trivial; and
- more seriously, it's not possible to make `--url` and the other flags achieve a consistent and safe SSL configuration. You can ask Marc for details.

knz commented:
Nevertheless, I still kept the fix to add warnings for ignored arguments and the `application_name` fix.

@mberhault please have a look since you provided guidance on this.

knz commented:
Review status: 6 of 10 files reviewed at latest revision, 5 unresolved discussions.

---

*[pkg/base/config.go, line 212 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23725#-L7R79TYl6laFkhDiPCR:-L7_KUiw8k2RfwP8AYdT:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b29c27b66fe2c468aa02db89ee62d715c1d12854/pkg/base/config.go#L212)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

s/of/if/
</blockquote></details>

Done.

---

*[pkg/cli/sql_util.go, line 471 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23725#-L7R90CWfhKbX6JN_AuJ:-L7_Krw0NqxVefNrXMsV:b-r5dte4) ([raw file](https://github.com/cockroachdb/cockroach/blob/b29c27b66fe2c468aa02db89ee62d715c1d12854/pkg/cli/sql_util.go#L471)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Per my other comment, later command line flags overriding earlier ones is a consistent experience. And it has precedent in something like `--database=foo --database=bar`. I think everyone would naturally expect the second flag to override the first. 
</blockquote></details>

Changed / simplified.

---

*[pkg/cli/user.go, line 41 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23725#-L7R7ZdYaBDh6pzKyOMR:-L7_KvS14Mgx5GQuy99A:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b29c27b66fe2c468aa02db89ee62d715c1d12854/pkg/cli/user.go#L41)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

I don't think we want the CLI subcommands to be this fine-grained. We should at least put all the `cockroach user` variants as one "application". Maybe `cockroach sql` gets its own "application" and all the rest are just `cockroach cli`.  
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23725)*
<!-- Sent from Reviewable.io -->

petermattis commented:
> more seriously, it's not possible to make --url and the other flags achieve a consistent and safe SSL configuration. You can ask Marc for details.

@mberhault For posterity, can you provide details?

knz commented:
From my understanding:
- the certificate manager expects all certificates to live 1) in the same directory 2) with specific file names. This is irreconcilable with custom paths in user-specified `sslcacert=` and `sslcert=` url options.
- #15318 highlights a practical aspect of the problem

petermattis commented:
> the certificate manager expects all certificates to live 1) in the same directory 2) with specific file names. This is irreconcilable with custom paths in user-specified sslcacert=, sslkey= and sslcert= url options.

Isn't this only a problem for the `start` command which doesn't take a `--url` argument? You could also invert my suggestion: rather than removing `sqlConn{DB,...}`, we remove `sqlConnURL` and change the handling of `--url` so that it extracts into the component flags. There are likely some practical difficulties with this and I don't want to hold up this PR (that was never my intention).

knz commented:
> You could also invert my suggestion: rather than removing sqlConn{DB,...}, we remove sqlConnURL and change the handling of --url so that it extracts into the component flags.

Yes that was my thought too, but that requires a rewrite of the CertificateManager. It's when I started to do that that Marc highlighted I might be overlooking some issues. I decided I didn't want to research this so close to the 2.0 hard cut deadline.

bdarnell commented:
I understand that TLS stuff is more complicated, but it seems like at least for `--database` it would be good to have the specific flag override the URL instead of being ignored. (especially because our startup output gives you a database-less URL, but we should be strongly encouraging the use of a connection-level database instead of the `set database`/`use` session variable. 

---

Reviewed 7 of 7 files at r3.
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23725#-:-L7_cqV4S6VpsdWKk4Fv:b-o8c3u7)*
<!-- Sent from Reviewable.io -->

knz commented:
@bdarnell handling `--database` in that way is trivial, I can do it (specifically, let it override the URL *if the URL didn't have a database to start with). would you be ok having `--database`  be the special case?

bdarnell commented:
`--database` is the main one, but I'd rather make `user`, `host`, and `port` work the same way and make the TLS flags the exception (and maybe we'll be able to make them override the URL in the future after some refactoring of the certificate manager)

mberhault commented:
I disagree. We should follow psql here: either you specify a `conninfo` (key/value pairs of settings, or URI) or you use flags, but you don't do a "merge" of both.

knz commented:
@mberhault I think it's OK to let arguments fill in any place in the URL _that was not set already_. The problem otherwise is that the URL might simply be incomplete.

knz commented:
Ok I have implemented the following behavior:

- if no `--url` is specified, the behavior is like previously, the URL is constructed from flags.
- if `--url` is specified:
  - anything already specified in `--url` prevails. If there's another flag specifying something that the URL already provides, a warning is printed (fixes #22376).
  - for each of the following components in the URL, if it's absent (not specified), then the flags are used to fill in:
    - the username (set up from `--user`, or `root` if no `--user`, if there was no user in the URL),
    - the SSL settings (set up from `--insecure`, `--certs-dir` etc if `sslmode` not already in URL,
    - the password (input from terminal if not already in URL),
    - the current database (set up from `--database`)

In addition, I added a check that was missing previously: if a user provides an explicit password for root in the URL but the connection is otherwise insecure, an error is produced.

PTAL

knz commented:
Note I have decided against handling host/port in that way because to me it makes no sense to have a URL without that information already.

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 8 of 11 files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23725#-:-L7a-ge7oIbllCtQarU1:bnfp4nl)*
<!-- Sent from Reviewable.io -->

knz commented:
Thanks for all the feedback. I learned a lot here.
